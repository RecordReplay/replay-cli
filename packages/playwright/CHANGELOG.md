# @replayio/playwright

## 3.1.3

### Patch Changes

- Updated dependencies [[`40a1ec8`](https://github.com/replayio/replay-cli/commit/40a1ec8a828b398605c3855746d675bea3090d0c)]:
  - @replayio/test-utils@3.0.3
  - @replayio/replay@0.22.8

## 3.1.2

### Patch Changes

- Updated dependencies [[`0488c9f`](https://github.com/replayio/replay-cli/commit/0488c9f9cbffe33b1a52b1109c7765802a0ed304)]:
  - @replayio/test-utils@3.0.2

## 3.1.1

### Patch Changes

- [#540](https://github.com/replayio/replay-cli/pull/540) [`a6f8f10`](https://github.com/replayio/replay-cli/commit/a6f8f105654b39c5c457dfac91c5169f0ba6cc04) Thanks [@Andarist](https://github.com/Andarist)! - Avoid reporting duplicated API errors

- Updated dependencies [[`a6f8f10`](https://github.com/replayio/replay-cli/commit/a6f8f105654b39c5c457dfac91c5169f0ba6cc04)]:
  - @replayio/test-utils@3.0.1

## 3.1.0

### Minor Changes

- [#513](https://github.com/replayio/replay-cli/pull/513) [`d56ebda`](https://github.com/replayio/replay-cli/commit/d56ebda0a761ee6cf531a1e86d0fa99f7f192df5) Thanks [@Andarist](https://github.com/Andarist)! - Use the current branch name as the default run title in non-CI runs

- [#536](https://github.com/replayio/replay-cli/pull/536) [`8343abe`](https://github.com/replayio/replay-cli/commit/8343abe8f74fc67ef4fd374d943b73fdcead5a5c) Thanks [@Andarist](https://github.com/Andarist)! - Correctly correlate recordings with tests utilizing `repeatEach`

### Patch Changes

- [#519](https://github.com/replayio/replay-cli/pull/519) [`34b1ba7`](https://github.com/replayio/replay-cli/commit/34b1ba705d5c6918333482707b5232fc8edf6170) Thanks [@Andarist](https://github.com/Andarist)! - Added new `upload` options:

  - `statusThreshold`: this accepts one of `'all'`, `'failed-and-flaky'` or `'failed'` and it's used to skip uploading recordings for test runs that don't satisfy the desired threshold (eg. it allows to filter out recordings of passed tests)
  - `minimizeUploads`: a boolean flag that helps to minimize the amount of uploaded recordings. With this flag a minimal set of recordings associated with a retried test is uploaded

- Updated dependencies [[`34b1ba7`](https://github.com/replayio/replay-cli/commit/34b1ba705d5c6918333482707b5232fc8edf6170), [`8343abe`](https://github.com/replayio/replay-cli/commit/8343abe8f74fc67ef4fd374d943b73fdcead5a5c), [`34b1ba7`](https://github.com/replayio/replay-cli/commit/34b1ba705d5c6918333482707b5232fc8edf6170), [`d56ebda`](https://github.com/replayio/replay-cli/commit/d56ebda0a761ee6cf531a1e86d0fa99f7f192df5), [`8343abe`](https://github.com/replayio/replay-cli/commit/8343abe8f74fc67ef4fd374d943b73fdcead5a5c)]:
  - @replayio/replay@0.22.7
  - @replayio/test-utils@3.0.0

## 3.0.3

### Patch Changes

- [#516](https://github.com/replayio/replay-cli/pull/516) [`c27e2af`](https://github.com/replayio/replay-cli/commit/c27e2afa983dab6668c90a7b4704ef42f4836ec7) Thanks [@Andarist](https://github.com/Andarist)! - Fixed a race condition that could cause some tests not being added correctly to a test run

- Updated dependencies [[`c27e2af`](https://github.com/replayio/replay-cli/commit/c27e2afa983dab6668c90a7b4704ef42f4836ec7)]:
  - @replayio/test-utils@2.1.2
  - @replayio/replay@0.22.6

## 3.0.2

### Patch Changes

- [#501](https://github.com/replayio/replay-cli/pull/501) [`e7c637c`](https://github.com/replayio/replay-cli/commit/e7c637ca95fc1ba649fd8cc87fc15059250f8ae1) Thanks [@Andarist](https://github.com/Andarist)! - Make `RECORD_REPLAY_ENABLE_ASSERTS` environment forwardable to the browser process

- [#502](https://github.com/replayio/replay-cli/pull/502) [`9122af4`](https://github.com/replayio/replay-cli/commit/9122af45618741cf2b222b461eddee016e38db43) Thanks [@Andarist](https://github.com/Andarist)! - Added a warning when the reporter gets used without the Replay browser

## 3.0.1

### Patch Changes

- [#488](https://github.com/replayio/replay-cli/pull/488) [`e6b79b4`](https://github.com/replayio/replay-cli/commit/e6b79b4821b894522bce0ea00f04e7d1ba6d7e3b) Thanks [@callingmedic911](https://github.com/callingmedic911)! - Skip Replay browser installation check on Windows
