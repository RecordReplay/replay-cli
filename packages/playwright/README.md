# @replayio/playwright

Provides utilities to support using [Replay](https://replay.io) with [Playwright](https://playwright.dev)

## Installation

`npm i @replayio/playwright`

## Usage

Use with [replayio/action-playwright](https://github.com/Replayio/action-playwright) to automatically upload replays of failed tests. [Check out our documentation here.](https://docs.replay.io/docs/recording-playwright-b62474b5aadd49e2b0c44a7580b0617e#4f9d9bb360974bf7942f8edae8dcd742)

### Exports

- `getExecutablePath(browserName: string)` - Returns the path to the replay browser for the given `browserName`: either `"chromium"` or `"firefox"`. If `browserName` isn't supported on the current platform, `undefined` is returned.
- `devices` - Object of configurations suitable for using with `@playwright/test`. Currently supports `"Replay Firefox"` and `"Replay Chromium"` configurations. If the configuration isn't supported on the current platform, a warning is emitted and the `executablePath` will be undefined.
- `getMetadataFilePath(workerIndex: number = 0)` - Returns the path of a worker-specific metadata file keyed by the `workerIndex`. The file path will be within the `RECORD_REPLAY_DIRECTORY`.

### Using standalone

If you are using `playwright` (rather than `@replayio/playwright`), you can configure it to use the Replay browser by passing in the `executablePath` to `launch()`.

> **Note:** For `firefox`, you must also pass the `RECORD_ALL_CONTENT` environment variable to start recording. This is not required for `chromium` which records all content by default.`

### Metadata

You can add metadata to your playwright recordings using either the `RECORD_REPLAY_METADATA` or `RECORD_REPLAY_METADATA_FILE` environment variable. If both are set, `RECORD_REPLAY_METADATA_FILE` takes precedence.

> Currently, this metadata is only available locally except for `title`

- `RECORD_REPLAY_METADATA_FILE` - The path to a file containing JSON-formatted metadata
- `RECORD_REPLAY_METADATA` - JSON-formatted metadata string

```js
const playwright = require("playwright");
const { getExecutablePath } = require("@replayio/playwright");

(async () => {
  const browser = await playwright.firefox.launch({
    headless: false,
    executablePath: getExecutablePath("firefox"),
    env: {
      RECORD_ALL_CONTENT: 1,
      RECORD_REPLAY_METADATA: JSON.stringify({
        title: "Screenshot of replay.io"
      })
    },
  });
  const page = await browser.newPage();
  await page.goto("https://replay.io");
  await page.screenshot({ path: "replay.png" });

  await page.close();
  await browser.close();
})();
```

### Using with `@playwright/test`

`@replayio/playwright` exports a `devices` object with configurations for both `"Replay Firefox"` and `"Replay Chromium"`. These can be added to your `playwright.config.js` to start recording your tests.

```js
// playwright.config.js
// @ts-check
const { devices } = require("@replayio/playwright");

/** @type {import('@playwright/test').PlaywrightTestConfig} */
const config = {
  forbidOnly: !!process.env.CI,
  retries: process.env.CI ? 2 : 0,
  use: {
    trace: "on-first-retry",
    defaultBrowserType: "firefox",
  },
  projects: [
    {
      name: "firefox",
      use: {
        ...devices["Replay Firefox"],
      },
    },
  ],
};

module.exports = config;
```

## Parallel runs on CI

If you have a large test suite, you might choose to split your test suite up and run them in parallel across multiple machines but still treat them as a single suite. By default, `@replayio/playwright` will generate a UUID for the suite and store it in the recording metadata under `test.run.id` but in this case each machine will have its own id. In order to link these independently ran tests together, you can generate your own UUID and set it in the `RECORD_REPLAY_TEST_RUN_ID` environment variable and it will be used instead of generating a value.
