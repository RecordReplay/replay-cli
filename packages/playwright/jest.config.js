/** @type {import('ts-jest/dist/types').InitialOptionsTsJest} */
module.exports = {
  testTimeout: 20000,
  preset: "ts-jest",
  testEnvironment: "node",
  testMatch: ["**/(*.)+(spec|test).[jt]s?(x)", "!**/fixtures-app/**"],
  // TODO: implement dependencyExtractor: the whole src directory and the whole fixture app should be added here
};
