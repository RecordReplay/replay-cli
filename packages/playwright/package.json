{
  "name": "@replayio/playwright",
  "version": "0.3.31",
  "description": "Configuration utilities for using the Replay browsers with playwright",
  "main": "src/index.js",
  "scripts": {
    "if:exists": "node -e \"require('fs').existsSync(process.argv[1]) ? require('child_process').spawnSync(process.argv[2], process.argv.slice(3), {shell: true, stdio: 'inherit'}) : 0\"",
    "if:source": "npm run if:exists -- src/index.ts",
    "if:dist": "npm run if:exists -- src/index.js",
    "prepare": "npm run if:source -- npm run build",
    "install": "npm run if:dist -- node ./bin/replayio-playwright first-run",
    "build": "rm -rf dist/ && tsc && chmod 755 dist/bin/* && cp package.json README.md dist/",
    "test": "echo \"Error: no test specified\"",
    "typecheck": "tsc --noEmit"
  },
  "bin": {
    "replayio-playwright": "./bin/replayio-playwright.js"
  },
  "publishConfig": {
    "directory": "dist"
  },
  "devDependencies": {
    "@playwright/test": "1.19.x",
    "@types/node": "^18.16.1",
    "@types/uuid": "^8.3.4",
    "typescript": "4.6.x"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/replayio/replay-cli.git"
  },
  "author": "",
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/replayio/replay-cli/issues"
  },
  "homepage": "https://github.com/replayio/replay-cli/blob/main/packages/playwright/README.md",
  "dependencies": {
    "@replayio/replay": "^0.12.11",
    "@replayio/test-utils": "^0.5.10",
    "uuid": "^8.3.2"
  },
  "peerDependencies": {
    "@playwright/test": "1.19.x"
  },
  "gitHead": "e5d1208c1691f442ef4aeba6a238ec0b76a03a9b"
}
