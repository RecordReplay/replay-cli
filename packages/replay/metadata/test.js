const {
  array,
  create,
  defaulted,
  enums,
  number,
  object,
  optional,
  string,
} = require("superstruct");

const VERSION = 1;

const versions = {
  1: object({
    file: optional(string()),
    path: optional(array(string())),
    result: enums(["passed", "failed", "timedOut"]),
    run: optional(string()),
    title: string(),
    version: defaulted(number(), () => 1),
  }),
};

function validate(metadata) {
  if (!metadata || !metadata.test) {
    throw new Error('Test metadata does not exist');
  }

  return init(metadata.test);
}

function init(data) {
  const version = data.version || VERSION;
  if (versions[version]) {
    return {
      test: create(data, versions[version]),
    };
  } else {
    throw new Error(`Test metadata version ${data.version} not supported`);
  }
}

module.exports = {
  validate,
  init
};
