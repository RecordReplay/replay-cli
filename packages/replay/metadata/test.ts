import { envString, firstEnvValueOf } from "./env";
import {
  array,
  create,
  defaulted,
  enums,
  number,
  object,
  optional,
  string,
  define,
  Struct,
} from "superstruct";
const isUuid = require("is-uuid");

import { UnstructuredMetadata } from "./types";

const VERSION = 1;

const versions: Record<number, Struct<any, any>> = {
  1: object({
    file: optional(envString("RECORD_REPLAY_METADATA_TEST_FILE")),
    path: optional(array(string())),
    result: defaulted(
      enums(["passed", "failed", "timedOut"]),
      firstEnvValueOf("RECORD_REPLAY_METADATA_TEST_RESULT")
    ),
    tests: optional(
      array(
        object({
          id: optional(string()),
          title: string(),
          path: optional(array(string())),
          relativePath: optional(string()),
          result: enums(["passed", "failed", "timedOut"]),
          error: optional(
            object({
              message: string(),
              line: optional(number()),
              column: optional(number()),
            })
          ),
        })
      )
    ),
    runner: optional(
      defaulted(
        object({
          name: optional(envString("RECORD_REPLAY_METADATA_TEST_RUNNER_NAME")),
          version: optional(envString("RECORD_REPLAY_METADATA_TEST_RUNNER_VERSION")),
        }),
        {}
      )
    ),
    run: optional(
      defaulted(
        object({
          id: defaulted(
            define("uuid", (v: any) => isUuid.v4(v)),
            firstEnvValueOf("RECORD_REPLAY_METADATA_TEST_RUN_ID", "RECORD_REPLAY_TEST_RUN_ID")
          ),
          title: optional(envString("RECORD_REPLAY_METADATA_TEST_RUN_TITLE")),
        }),
        {}
      )
    ),
    title: envString("RECORD_REPLAY_METADATA_TEST_TITLE"),
    version: defaulted(number(), () => 1),
  }),
};

function validate(metadata: { test: UnstructuredMetadata }) {
  if (!metadata || !metadata.test) {
    throw new Error("Test metadata does not exist");
  }

  return init(metadata.test);
}

function init(data: UnstructuredMetadata = {}) {
  const version = typeof data.version === "number" ? data.version : VERSION;
  if (versions[version]) {
    return {
      test: create(data, versions[version]),
    };
  } else {
    throw new Error(`Test metadata version ${data.version} not supported`);
  }
}

export { validate, init };
