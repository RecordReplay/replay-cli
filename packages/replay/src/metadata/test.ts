export * from "@replay-cli/shared/recording/metadata/legacy/test/index";
