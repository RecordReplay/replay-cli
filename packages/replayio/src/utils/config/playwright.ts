import { DeclarationKind, ExpressionKind } from "ast-types/gen/kinds";
import { execSync } from "child_process";
import { existsSync, readFileSync, writeFileSync } from "fs";
import chalk from "chalk";
import jscodeshift, { ArrayExpression, ObjectExpression, Property } from "jscodeshift";
import path from "path";
import { th } from "date-fns/locale";

const j = jscodeshift.withParser("ts");

export function findPlaywrightConfig() {
  const possibleNames = ["playwright.config.js", "playwright.config.ts"];
  for (const name of possibleNames) {
    const fullPath = path.join(process.cwd(), name);
    if (existsSync(fullPath)) {
      return fullPath;
    }
  }
  return null;
}

export function setupReplayConfig(source: string) {
  const root = j(source);

  function ensureReplayImports() {
    const hasReplayImport =
      root
        .find(j.ImportDeclaration, {
          source: { value: "@replayio/playwright" },
        })
        .size() > 0;

    if (hasReplayImport) {
      console.log(chalk.yellow("Replay imports already exist in the playwright.config.ts file."));
      throw new Error("Replay imports already exist in the playwright.config.ts file.");
    }

    const isESModule = root.find(j.ImportDeclaration).size() > 0;

    if (isESModule) {
      root
        .find(j.ImportDeclaration)
        .at(-1)
        .insertAfter("import { devices as replayDevices } from '@replayio/playwright';");
    } else {
      root
        .find(j.VariableDeclaration)
        .at(-1)
        .insertAfter("const replayDevices = require('@replayio/playwright').devices;");
    }
  }

  function extractConfigObject(node: DeclarationKind | ExpressionKind) {
    // Directly exported object. Not sure if this is a valid case.
    if (node.type === "ObjectExpression") {
      return node;
    }
    // Call to defineConfig
    if (
      node.type === "CallExpression" &&
      node.callee.type === "Identifier" &&
      node.callee.name === "defineConfig" &&
      node.arguments.length > 0
    ) {
      return node.arguments[0].type === "ObjectExpression" ? node.arguments[0] : null;
    }
    return null;
  }

  function extractConfigObjectFromRequire(node: DeclarationKind | ExpressionKind) {
    if (
      node.type === "CallExpression" &&
      node.callee.type === "Identifier" &&
      node.callee.name === "require" &&
      node.arguments.length === 1 &&
      node.arguments[0].type === "Literal"
    ) {
      // Case where configuration is required from another module
      // Theoretically, we'd need to read and parse the required file. But we'll ignore that for now.
      console.log(chalk.yellow("Playwright configuration is required from another module. "));
    }
    return null;
  }

  function updateConfigObject() {
    root.find(j.ExportDefaultDeclaration).forEach(path => {
      const configObject =
        extractConfigObject(path.value.declaration) ||
        extractConfigObjectFromRequire(path.value.declaration);

      if (configObject) {
        modifyConfigObject(configObject);
      }
    });

    // Handle module.exports for CommonJS explicitly
    root
      .find(j.AssignmentExpression, {
        left: {
          type: "MemberExpression",
          object: { type: "Identifier", name: "module" },
          property: { type: "Identifier", name: "exports" },
        },
      })
      .forEach(path => {
        const configObject =
          extractConfigObject(path.value.right) || extractConfigObjectFromRequire(path.value.right);

        if (configObject) {
          modifyConfigObject(configObject);
        }
      });

    function modifyConfigObject(configObject: ObjectExpression) {
      if (configObject.type === "ObjectExpression") {
        let reporterPropertyIndex = configObject.properties.findIndex(
          (prop: any) =>
            prop.type === "ObjectProperty" &&
            prop.key.type === "Identifier" &&
            prop.key.name === "reporter"
        );

        if (reporterPropertyIndex === -1) {
          const newReporterProperty = j.property(
            "init",
            j.identifier("reporter"),
            j.arrayExpression([])
          );
          configObject.properties.push(newReporterProperty);
          reporterPropertyIndex = configObject.properties.length - 1;
        }

        const reporterProperty = configObject.properties[reporterPropertyIndex] as Property;
        if (reporterProperty.value.type !== "ArrayExpression") {
          const newValue = reporterProperty.value;
          reporterProperty.value = j.arrayExpression([newValue] as any);
        }

        reporterProperty.value.elements.push(
          j.arrayExpression([
            j.literal("@replayio/playwright/reporter"),
            j.objectExpression([
              j.property(
                "init",
                j.identifier("apiKey"),
                j.memberExpression(
                  j.memberExpression(j.identifier("process"), j.identifier("env")),
                  j.identifier("REPLAY_API_KEY")
                )
              ),
              j.property("init", j.identifier("upload"), j.literal(true)),
            ]),
          ])
        );

        let projectsPropertyIndex = configObject.properties.findIndex(
          (prop: any) =>
            prop.type === "ObjectProperty" &&
            prop.key.type === "Identifier" &&
            prop.key.name === "projects"
        );

        if (projectsPropertyIndex === -1) {
          const newProjectsProperty = j.property(
            "init",
            j.identifier("projects"),
            j.arrayExpression([])
          );
          configObject.properties.push(newProjectsProperty);
          projectsPropertyIndex = configObject.properties.length - 1;
        }

        const projectsProperty = configObject.properties[projectsPropertyIndex] as Property;
        const projectsValueArray = projectsProperty.value as ArrayExpression;
        projectsValueArray.elements.push(
          j.objectExpression([
            j.property("init", j.identifier("name"), j.literal("replay-chromium")),
            j.property(
              "init",
              j.identifier("use"),
              j.objectExpression([
                j.spreadElement(
                  j.memberExpression(
                    j.identifier("replayDevices"),
                    j.literal("Replay Chromium"),
                    true
                  )
                ),
              ])
            ),
          ])
        );
      }
    }
  }

  ensureReplayImports();
  updateConfigObject();

  return root.toSource({ parser: "ts" });
}

export function setupReplayInPlaywrightConfig(configPath: string) {
  const source = readFileSync(configPath, { encoding: "utf8" });

  const updatedSource = setupReplayConfig(source);

  writeFileSync(configPath, updatedSource, { encoding: "utf8" });
}

export function installReplayDependencies(projectDir: string) {
  let command: string;

  if (existsSync(path.join(projectDir, "package-lock.json"))) {
    command = `npm install --save-dev @replayio/playwright`;
  } else if (existsSync(path.join(projectDir, "pnpm-lock.yaml"))) {
    command = `pnpm add --save-dev @replayio/playwright`;
  } else if (existsSync(path.join(projectDir, "yarn.lock"))) {
    command = `yarn add --dev @replayio/playwright`;
  } else if (existsSync(path.join(projectDir, "bun.lockb"))) {
    command = `bun add --dev @replayio/playwright`;
  } else {
    console.error("No supported package manager found.");
    throw new Error("No supported package manager found.");
  }

  execSync(command, { cwd: projectDir });
}
