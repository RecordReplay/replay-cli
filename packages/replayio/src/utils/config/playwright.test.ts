import { setupReplayConfig } from "./playwright";

describe("setupReplayConfig", () => {
  it("should add Replay configurations correctly for ES module syntax", () => {
    const source = `
import { defineConfig, devices } from '@playwright/test'

export default defineConfig({
  testDir: './tests',

  /* Run tests in files in parallel */
  fullyParallel: true,

  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,

  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,

  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,

  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [['line'], ['html', { open: 'never' }]],

  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    trace: 'on-first-retry'
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] }
    }
  ],

  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'yarn dev',
    url: 'http://127.0.0.1:3000',
    reuseExistingServer: !process.env.CI
  }
})
`;
    const transformedSource = setupReplayConfig(source);
    expect(transformedSource).toMatchSnapshot();
  });

  it("should add Replay configurations correctly for CommonJS syntax", () => {
    const source = `
const { defineConfig, devices } = require('@playwright/test');

module.exports = defineConfig({
  testDir: './tests',

  /* Run tests in files in parallel */
  fullyParallel: true,

  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,

  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,

  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,

  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [['line'], ['html', { open: 'never' }]],

  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    trace: 'on-first-retry'
  },

  /* Configure projects for major browsers */
  projects: [
    {
      name: 'chromium',
      use: { ...devices['Desktop Chrome'] }
    }
  ],

  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'yarn dev',
    url: 'http://127.0.0.1:3000',
    reuseExistingServer: !process.env.CI
  }
});
`;
    const transformedSource = setupReplayConfig(source);
    expect(transformedSource).toMatchSnapshot();
  });

  it("should handle existing Replay configurations gracefully", () => {
    const source = `
import { devices as replayDevices } from '@replayio/playwright';
import { test, expect } from '@playwright/test';

export default defineConfig({
  reporter: [
    ["@replayio/playwright/reporter", { apiKey: process.env.REPLAY_API_KEY, upload: true }],
    // Other reporters...
  ],
  projects: [
    {
      name: 'replay-chromium',
      use: { ...replayDevices["Replay Chromium"] },
    },
    // Other projects...
  ],
});
`;
    expect(() => setupReplayConfig(source)).toThrow(
      "Replay imports already exist in the playwright.config.ts file."
    );
  });
});
