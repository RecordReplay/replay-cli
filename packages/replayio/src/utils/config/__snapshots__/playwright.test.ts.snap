// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`setupReplayConfig should add Replay configurations correctly for CommonJS syntax 1`] = `
"
const { defineConfig, devices } = require('@playwright/test');

const replayDevices = require('@replayio/playwright').devices;

module.exports = defineConfig({
  testDir: './tests',

  /* Run tests in files in parallel */
  fullyParallel: true,

  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,

  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,

  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,

  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [['line'], ['html', { open: 'never' }], [\\"@replayio/playwright/reporter\\", {
    apiKey: process.env.REPLAY_API_KEY,
    upload: true
  }]],

  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    trace: 'on-first-retry'
  },

  /* Configure projects for major browsers */
  projects: [{
    name: 'chromium',
    use: { ...devices['Desktop Chrome'] }
  }, {
    name: \\"replay-chromium\\",

    use: {
      ...replayDevices[\\"Replay Chromium\\"]
    }
  }],

  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'yarn dev',
    url: 'http://127.0.0.1:3000',
    reuseExistingServer: !process.env.CI
  }
});
"
`;

exports[`setupReplayConfig should add Replay configurations correctly for ES module syntax 1`] = `
"
import { defineConfig, devices } from '@playwright/test'

import { devices as replayDevices } from '@replayio/playwright';

export default defineConfig({
  testDir: './tests',

  /* Run tests in files in parallel */
  fullyParallel: true,

  /* Fail the build on CI if you accidentally left test.only in the source code. */
  forbidOnly: !!process.env.CI,

  /* Retry on CI only */
  retries: process.env.CI ? 2 : 0,

  /* Opt out of parallel tests on CI. */
  workers: process.env.CI ? 1 : undefined,

  /* Reporter to use. See https://playwright.dev/docs/test-reporters */
  reporter: [['line'], ['html', { open: 'never' }], [\\"@replayio/playwright/reporter\\", {
    apiKey: process.env.REPLAY_API_KEY,
    upload: true
  }]],

  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    trace: 'on-first-retry'
  },

  /* Configure projects for major browsers */
  projects: [{
    name: 'chromium',
    use: { ...devices['Desktop Chrome'] }
  }, {
    name: \\"replay-chromium\\",

    use: {
      ...replayDevices[\\"Replay Chromium\\"]
    }
  }],

  /* Run your local dev server before starting the tests */
  webServer: {
    command: 'yarn dev',
    url: 'http://127.0.0.1:3000',
    reuseExistingServer: !process.env.CI
  }
})
"
`;
